### PR TITLE
drivers: can: mcan: use per-instance message RAM configuration

### LINKED ISSUE #42030: can: "bosch,m-can-base": Warning "missing or empty reg/ranges property"
**Describe the bug**
When building (any application) one of the boards that include can node with compatible "bosch,m-can-base",
the following warning is generated:
```
~/zephyrproject/zephyr/samples/synchronization/build/nucleo_h723zg/zephyr/zephyr.dts:2708.7-2734.5: Warning (simple_bus_reg): /soc/can: missing or empty reg/ranges property
```

**To Reproduce**
`west build -b nucleo_h723zg samples/hello_world`

**Expected behavior**
No warning generated

**Impact**
Annoyance

### LINKED ISSUE #53417: CAN-FD / MCAN driver: Possible variable overflow for some MCUs
**Describe the bug**

The CAN-FD driver stores different flags for the filters as a bitset in the following variables:

https://github.com/zephyrproject-rtos/zephyr/blob/e4fcb32451c587a3e4ba7f8bf3fc602b16f9652b/drivers/can/can_mcan.h#L181-L184

These variables should have as many bits as the maximum number of filters of this type.

For the NXP LPC55Sxx the `uint8_t` is not sufficient for the number of extended filters:

https://github.com/zephyrproject-rtos/zephyr/blob/e4fcb32451c587a3e4ba7f8bf3fc602b16f9652b/dts/arm/nxp/nxp_lpc55S1x_common.dtsi#L216

The variable size should be increased to `uint16_t`.

**Impact**

Other flags would be overwritten if an application actually uses more than 8 extended ID filters.

The overflow has no impact on device security, as far as I can see. However, it could lead to unexpected behavior of the CAN driver.

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Switch the Bosch M_CAN devicetree binding to use a bosch,mram-cfg property for specifying the memory layout of the Bosch M_CAN Message RAM. This is  identical to the Linux kernel devicetree binding for Bosch M_CAN IP core based CAN controllers. This introduces an offset cell which can be used for controllers with shared Message RAM between Bosch M_CAN instances.

Restructure the Bosch M_CAN driver backend to use per-instance Message RAM configuration. This removes the need for a common, artificial "can" devicetree node for SoCs with multiple Bosch M_CAN-based CAN controllers and allows for per-instance configuration of the number of e.g. standard (11-bit) and extended (29-bit) filter elements.
    
As part of the restructure, software handling of CAN filter flags was moved from per-flags bitfields to per-filter bitfields, solving an issue when using more than 32 standard (11-bit) filter elements or more than 16 extended (29-bit) filter elements.

Now there is no longer a need for iterating all the Bosch M_CAN filter elements in  Message RAM in order to find a free filter as the driver already keeps track of assigned filters.
    
Fixes: #42030, #53417

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
I would appreciate testing this on the following platforms, for which I have only compile-time tested these changes:
- ST STM32H7 (@FRASTM?)
- Atmel SAM0 (@bennys, @nandojve)
- Atmel SAM (@nandojve)

**Comment by gramsay0** [comment]
Great work! `can_mcan_ops` will make MMIO much easier too

File: drivers/can/can_mcan.c [hunk 1]
```diff
@@ -76,6 +48,27 @@ int can_mcan_write_reg(const struct device *dev, uint16_t reg, uint32_t val)
 	return err;
 }
 
+int can_mcan_read_mram(const struct device *dev, uint16_t offset, void *dst, size_t len)
```
Comment (gramsay0) [inline drivers/can/can_mcan.c:51]: Could these functions be static?

File: drivers/can/can_mcux_mcan.c [hunk 2]
```diff
@@ -170,13 +190,26 @@ static const struct can_driver_api mcux_mcan_driver_api = {
 #endif /* CONFIG_CAN_FD_MODE */
 };
 
+struct can_mcan_ops mcux_mcan_ops = {
```
Comment (gramsay0) [inline drivers/can/can_mcux_mcan.c:193]: Could this be const?
(same in other drivers)

File: drivers/can/can_mcan.c (hunk 1 above)
Comment (henrikbrixandersen) [reply drivers/can/can_mcan.c:51]: They cannot, unfortunately. They are exposed in the `can_mcan.h` header for use in out-of-tree driver implementations, where the frontend driver (may) need to access Message RAM directly. One could argue that the frontend driver could just use its private MRAM access functions, but the parameters there are different.

It's just one function call deeper by not inlining, I think it is okay.

File: drivers/can/can_mcux_mcan.c (hunk 2 above)
Comment (henrikbrixandersen) [reply drivers/can/can_mcux_mcan.c]: Yes! That was the intention too, fixed. Thanks.

File: drivers/can/can_mcan.c (hunk 1 above)
Comment (gramsay0) [reply drivers/can/can_mcan.c:51]: I agree, that sounds reasonable

File: drivers/can/can_mcux_mcan.c (hunk 2 above)
Comment (gramsay0) [reply drivers/can/can_mcux_mcan.c]: Could you add `static` too

Comment (henrikbrixandersen) [reply drivers/can/can_mcux_mcan.c]: Just did, thanks.

**Comment by nandojve** [comment]
> I would appreciate testing this on the following platforms, for which I have only compile-time tested these changes:
> 
> * ST STM32H7 (@FRASTM?)
> * Atmel SAM0 (@benys? @nandojve?)
> * Atmel SAM (@nandojve?)

Hi @henrikbrixandersen ,

`west build -b sam_v71_xult samples/drivers/can/counter` seems OK.
I don't have any SAM0 board with CAN to test.

```shell
*** Booting Zephyr OS build zephyr-v3.3.0-4641-g3c43f12e05db ***
Change LED filter ID: 0
Finished init.
Counter filter id: 28


uart:~$ Counter received: 0
Counter received: 1
Counter received: 2
```

Is this enough to validate or do you prefer any other sample/test?

**Comment by henrikbrixandersen** [comment]
> Is this enough to validate or do you prefer any other sample/test?

@nandojve Thank you! That's all I needed.

**Review by nandojve — DISMISSED** [review]

**Comment by nandojve** [comment]
> > Is this enough to validate or do you prefer any other sample/test?
>
> @nandojve Thank you! That's all I needed.

I @benys can check SAM0.

**Comment by benys** [comment]
Ok:) i can check it with c21 but at Monday

**Review by gramsay0 — DISMISSED** [review]
I noticed these minor issues (unrelated to your changes), possibly could sneak the fixes in here...

drivers/can/can_mcan.c:855 - Format string warning, should be `%zu`: `LOG_DBG("Sending %d bytes`
drivers/can/can_sam0.c:45 - Unused variable: `uint32_t bits = 0U;`

File: drivers/can/can_mcan.c (hunk 1 above)
Comment (alexanderwachter) [inline drivers/can/can_mcan.c:51]: Woulit be an option to put them in the header and mark them inline?

**Comment by henrikbrixandersen** [comment]
> I noticed these minor issues (unrelated to your changes), possibly could sneak the fixes in here...
> 
> drivers/can/can_mcan.c:855 - Format string warning, should be `%zu`: `LOG_DBG("Sending %d bytes` drivers/can/can_sam0.c:45 - Unused variable: `uint32_t bits = 0U;`

Thank you for the thorough review. I'll address those two issues in separate PRs.

File: drivers/can/can_mcan.c (hunk 1 above)
Comment (henrikbrixandersen) [reply drivers/can/can_mcan.c]: Good idea. I have moved the implementations to the header file and inlined them.

File: drivers/can/can_mcan.h [hunk 3]
```diff
@@ -700,6 +1278,116 @@ static inline int can_mcan_sys_write_reg(mm_reg_t base, uint16_t reg, uint32_t v
...
+ *        must be 32-bit aligned.
+ * @param offset Offset from the start of the Message RAM for the given Bosch M_CAN instance. The
+ *        offset must be 32-bit aligned.
+ * @param src Source for the data to be written. The source address must be 32-bit aligned.
+ * @param len Number of bytes to write. Must be a multiple of 4.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General input/output error.
+ */
+static inline int can_mcan_sys_write_mram(mem_addr_t base, uint16_t offset, const void *src,
```
Comment (alexanderwachter) [inline drivers/can/can_mcan.h:1334]: Just out of curiosity, why 'static' ?
Redefinition is prevented by 'inline'. For me 'static' is the indicator that it is only used within this file. Sure, the preprocessor include let the function appear in the file, but it somehow feels wrong.
Am I missing some other quirk from static?

Comment (henrikbrixandersen) [reply drivers/can/can_mcan.h:1334]: Please see https://stackoverflow.com/questions/21835664/why-declare-a-c-function-as-static-inline

Comment (alexanderwachter) [reply drivers/can/can_mcan.h:1334]: Ahhh, we want to have the possibility that the compiler can throw away the global visible function and inline the ones used.

Comment (henrikbrixandersen) [reply drivers/can/can_mcan.h:1334]: Exactly.

File: drivers/can/can_mcan.h [hunk 4]
```diff
@@ -605,19 +1133,64 @@ struct can_mcan_config {
...
+
 /**
  * @brief Static initializer for @p can_mcan_config struct
  *
  * @param node_id Devicetree node identifier
  * @param _custom Pointer to custom driver frontend configuration structure
- * @param _read_reg Driver frontend Bosch M_CAN register read function
- * @param _write_reg Driver frontend Bosch M_CAN register write function
+ * @param _ops Pointer to front-end @a can_mcan_ops
+ * @param _ops Pointer to front-end @a can_mcan_callbacks
```
Comment (gramsay0) [inline drivers/can/can_mcan.h:1182]: Typo `_ops` -> `_cbs`
(Same for `CAN_MCAN_DT_CONFIG_INST_GET`)

File: drivers/can/can_mcan.h [hunk 5]
```diff
@@ -651,21 +1229,21 @@ struct can_mcan_config {
...
  * @see CAN_MCAN_DT_CONFIG_GET()
  */
-#define CAN_MCAN_DT_CONFIG_INST_GET(inst, _custom, _read_reg, _write_reg)                          \
-	CAN_MCAN_DT_CONFIG_GET(DT_DRV_INST(inst), _custom, _read_reg, _write_reg)
+#define CAN_MCAN_DT_CONFIG_INST_GET(inst, _custom, _ops, _cbs)                                     \
+	CAN_MCAN_DT_CONFIG_GET(DT_DRV_INST(inst), _custom, _ops, _cbs)
 
 /**
  * @brief Initializer for a @a can_mcan_data struct
  * @param _msg_ram Pointer to message RAM structure
```
Comment (gramsay0) [inline drivers/can/can_mcan.h:1241]: Remove `_msg_ram` parameter from docs

File: drivers/can/can_mcan.h (hunk 4 above)
Comment (henrikbrixandersen) [reply drivers/can/can_mcan.h]: Thanks, fixed.

File: drivers/can/can_mcan.h (hunk 5 above)
Comment (henrikbrixandersen) [reply drivers/can/can_mcan.h]: Fixed!

**Review by gramsay0 — APPROVED** [review]
@henrikbrixandersen FYI I got CAN working on a TI AM62x today (it also uses MCAN). I worked off this branch as it made it easier to develop 👍

**Comment by benys** [comment]
@henrikbrixandersen I run your PR on my C21 board. It works for me - messages are handled :).

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen I run your PR on my C21 board. It works for me - messages are handled :).

Thanks!

**Comment by FRASTM** [comment]
> I would appreciate testing this on the following platforms, for which I have only compile-time tested these changes:
> 
> * ST STM32H7 (@FRASTM?)

Successfully Tested the samples/drivers/can/counter on the stm32h7b3i_dk
```
*** Booting Zephyr OS build zephyr-v3.3.0-4502-ged58fb7cb2b1 ***
Change LED filter ID: 0
Finished init.
Counter filter id: 28


uart:~$ Counter received: 0
Counter received: 1
Counter received: 2
Counter received: 3
Counter received: 4
Counter received: 5
Counter received: 6
Counter received: 7
```

Note that it does not fix the https://github.com/zephyrproject-rtos/zephyr/issues/54937 on the nucleo_g474re

**Comment by henrikbrixandersen** [comment]
> Successfully Tested the samples/drivers/can/counter on the stm32h7b3i_dk

Thanks for testing!

> also PASSED with nucleo_g474re
> 
> Note that it does not fix the #54937 on the nucleo_g474re

Yeah, I know, thanks. I have that board myself. I still have not found the root cause for #54937 but I do not think it is a CAN driver issue, as all other CAN tests pass on this board. I will handle this issue separately.

**Comment by henrikbrixandersen** [comment]
All: Thank you for assisting with reviews and testing. Please revisit and (re-)approve so this PR can get merged for v3.4.0.

**Review by nandojve — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**